### PR TITLE
[sled-agent] Avoid causing UUID conflicts

### PR DESCRIPTION
**Pull request by smklein**
- Avoids overwriting the value of "dataset UUID" when creating datasets from `omicron_zones_ensure`. Instead, don't set any dataset UUID, which lets subsequent calls to `datasets_ensure` set the right value here. 

Fixes https://github.com/oxidecomputer/omicron/issues/7265

### CODE REVIEW

File: sled-storage/src/manager.rs [hunk 1]
```diff
@@ -526,7 +524,7 @@ impl StorageHandle {
     // and ask for the set of all datasets from Nexus.
     pub async fn upsert_filesystem(
```
Comment (smklein) [inline sled-storage/src/manager.rs:525]: FWIW, this is the **only** call to `upsert_filesystem`. Once we remove the invocation in `omicron_zones_ensure`, we can remove this function (and this "optional dataset UUID" argument) altogether.

File: sled-agent/src/sled_agent.rs [hunk 2]
```diff
@@ -962,12 +962,20 @@ impl SledAgent {
...
+            //
+            // However, we need to ensure that all blueprints have datasets
+            // within them before we can remove this back-fill.
+            //
+            // Therefore, we do something hairy here: We ensure the filesystem
+            // exists, but don't specify any dataset UUID value.
+            //
+            // This means that:
+            // - If the dataset exists and has a UUID, this will be a no-op
+            // - If the dataset doesn't exist, it'll be created
```
Comment (jgallagher) [inline sled-agent/src/sled_agent.rs:975]: Maybe add something like

```suggestion
            // - If the dataset doesn't exist, it'll be created without its oxide:uuid zfs property set
```

File: sled-storage/src/manager.rs [hunk 3]
```diff
@@ -1499,27 +1498,9 @@ impl StorageManager {
             zoned,
             encryption_details,
             size_details,
-            id: Some(DatasetUuid::from_untyped_uuid(request.dataset_id)),
+            id: request.dataset_id,
             additional_options: None,
         })?;
-        // Ensure the dataset has a usable UUID.
```
Comment (jgallagher) [inline sled-storage/src/manager.rs:1505]: I'm still a little weirded out by this block of code. Was it strictly unnecessary before? If there's any chance it wasn't, should we keep it but inside an `if let Some(dataset_id) { ... }` block?

File: sled-storage/src/manager.rs [hunk 4]
```diff
@@ -2005,16 +1985,39 @@ mod tests {
             .expect("Ensuring disks should work after key manager is ready");
         assert!(!result.has_error(), "{:?}", result);
 
-        // Create a filesystem on the newly formatted U.2
-        let dataset_id = Uuid::new_v4();
+        // Create a filesystem on the newly formatted U.2.
+        //
+        // We can call "upsert_filesystem" both with and without a UUID.
```
Comment (jgallagher) [inline sled-storage/src/manager.rs:1990]: Would it make sense to add a test for calling with `None` and seeing the dataset come back without an ID?

Comment (smklein) [reply sled-storage/src/manager.rs:1990]: Sure, added a new test: `upsert_filesystem_no_uuid`

File: sled-agent/src/sled_agent.rs (hunk 2 above)
Comment (smklein) [reply sled-agent/src/sled_agent.rs]: updated

File: sled-storage/src/manager.rs (hunk 3 above)
Comment (smklein) [reply sled-storage/src/manager.rs:1505]: I *think* it's a historical artifact -- I'd prefer to remove it if we can, because I think sled agent is plagued by too many of these "two ways to set a value" paths already, and I'd prefer to reduce it to one.

From 256dca4de51063689ea187c1c3f2a931b19fd3cf (2021) - https://github.com/oxidecomputer/omicron/blob/256dca4de51063689ea187c1c3f2a931b19fd3cf/sled-agent/src/storage_manager.rs#L232-L246 This is one of the earliest forms of "ensure filesystem, then set a value".

From ccc28fe4c857c08599f5d9d6eff6ecfcaa298eb6 (2023) - this was renamed, but still had the structure of "sled agent calls `Zfs::ensure_filesystem`, and **then** sets the UUID: https://github.com/oxidecomputer/omicron/blob/ccc28fe4c857c08599f5d9d6eff6ecfcaa298eb6/sled-agent/src/storage_manager.rs#L269-L298 

From de8bc93309f9eab8c431328065ea6f3d5bd0f34f (last week) -  `Zfs::ensure_filesystem` (later renamed to `ensure_dataset`) started being able to explicitly set properties, as part of the optimization to avoid calling `zfs set` individually for each property.

https://github.com/oxidecomputer/omicron/blob/de8bc93309f9eab8c431328065ea6f3d5bd0f34f/illumos-utils/src/zfs.rs#L550-L557

This was added recently, because (using dtrace) I noticed the callstack was basically:

```
zfs set <size properties , one at a time>
zfs get uuid
zfs set uuid
```

And by moving all properties into `Zfs::ensure_filesystem`, this could be a single call to `get/set`.

All this is to say:

- After de8bc93309f9eab8c431328065ea6f3d5bd0f34f, any property-managing code **after** calling `Zfs::ensure_filesystem` should not be necessary. If the UUID was supplied, it should have already been checked/set at this point.
- I'd prefer to remove this block of code, rather than diverge the code which mutates dataset UUIDs further.

Does that sound reasonable?

Comment (jgallagher) [reply sled-storage/src/manager.rs:1505]: Hah! Yes, I was doing the same digging and came to the same conclusion (i.e., #7236 made this code unnecessary). :+1: on the removal.

**Review by jgallagher — APPROVED** [review]
Changes LGTM; happy to go ahead and approve so this is ready to go once the testing from @iliana / @leftwo comes back.

**Review by iliana — APPROVED** [review]
Blueprint execution now works on a newly-initialized rack (rel/v12 + this PR backport).
